desc/magnetic_fields.py

```diff
@@ -1819,9 +1836,875 @@ def from_surface(
+                        this_minSeparation2 = jnp.min(d)
+                        if this_minSeparation2 < minSeparation2:
+                            minSeparation2 = this_minSeparation2
+
+                print(f"Minimum coil-coil separation: {minSeparation2*1000:3.2f} mm")
+            return contour_X, contour_Y, contour_Z
+
+        contour_X, contour_Y, contour_Z = find_XYZ_points(
+            contour_theta,
+            contour_zeta,
+            winding_surf,
+            find_min_dist=True,
+        )
+        ################################################################
+        # Create CoilSet object
+        ################################################################
+        # local imports to avoid circular imports
+        from desc.coils import MixedCoilSet, SplineXYZCoil
+
+        coils = []
+        for j in range(len(contour_X)):
+            if not jnp.isclose(helicity, 0):
+                # helical coils
+                # make sure that the sign of the coil current is correct
+                # by dotting K with the vector along the contour
+                contour_vector = jnp.array(
+                    [
+                        contour_X[j][1] - contour_X[j][0],
+                        contour_Y[j][1] - contour_Y[j][0],
+                        contour_Z[j][1] - contour_Z[j][0],
+                    ]
+                )
+                K = surface_current_field.compute(
+                    "K",
+                    grid=Grid(
+                        jnp.array([[0, contour_theta[j][0], contour_zeta[j][0]]])
+                    ),
+                    basis="xyz",
+                )["K"]
+                current_sign = jnp.sign(jnp.dot(contour_vector, K[0, :]))
+                thisCurrent = current_sign * coil_current
+            else:
+                # modular coils
+                # make sure that the sign of the coil current is correct
+                # don't need to dot with K here because we know the direction
+                # based off the direction of the theta contour and sign of G
+                # (extra negative sign because a positive G -> negative toroidal B
+                # but we always have a right-handed coord system, and so current flowing
+                # in positive poloidal direction creates a positive toroidal B)
+                current_sign = -sign_of_theta_contours * jnp.sign(net_poloidal_current)
+                thisCurrent = jnp.abs(coil_current) * current_sign
+            coils.append(
+                SplineXYZCoil(
+                    thisCurrent,
+                    jnp.append(contour_X[j][0::step], contour_X[j][0]),
+                    jnp.append(contour_Y[j][0::step], contour_Y[j][0]),
+                    jnp.append(contour_Z[j][0::step], contour_Z[j][0]),
+                    method=spline_method,
+                )
+            )
+
+        final_coilset = MixedCoilSet(*coils)
```

Comment:
More efficient if it's a regular coilset, mixed is only if there are different types/resolutions